File: Algorithm/排序/插入排序/insertion_sort.py
# -*- coding:utf-8 -*-
def insert_sort(raw_list):
    length = len(raw_list)
    for i in range(1, length):
        temp = raw_list[i]
        #j从i-1 到 0倒序
        for j in range(i-1,-1,-1):
            if(raw_list[j]<=temp):
                break
            if(raw_list[j]>temp):
                raw_list[j],raw_list[j+1] = temp,raw_list[j]
    return raw_list

#测试
print("插入排序测试")
print("排序前: ")
testbuf= [100, 7, 9, -2, 20, -50]
print(testbuf)
sorted_list = insert_sort(testbuf)
print("排序后: ")
print(sorted_list)

File: Algorithm/排序/插入排序/insertion_sort.c
#include <stdio.h>

void insert_sort(int *buf, int len)
{
    int temp;
    for(int i = 1; i < len; i++)
    {
        if(buf[i] < buf[i-1])
        {
            temp = buf[i];
            int j = i -1;
            for(; j >= 0 && buf[j] > temp; j--)
            {
                buf[j+1] = buf[j];
            }
            buf[j+1] = temp;
        }
    }
}

int testbuf[] = {100, 7, 9, -2, 20, -50};

int main(void)
{
    int testlen = sizeof(testbuf)/sizeof(int);
    printf("插入排序测试\n");
    printf("排序前: \n");
    for(int i = 0; i < testlen; i++)
    {
        printf("%d ", testbuf[i]);
    }
    printf("\n");
    insert_sort(testbuf, testlen);
    printf("排序后: \n");
    for(int i = 0; i < testlen; i++)
    {
        printf("%d ", testbuf[i]);
    }
    printf("\n");
}